wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceService.java

```diff
@@ -952,11 +963,87 @@ private WorkspaceDto asDtoWithLinksAndToken(WorkspaceImpl workspace) throws Serv
       } catch (MachineTokenException e) {
         throw new ServerException(e.getMessage(), e);
       }
+
+      fixRuntimeCommandsPreviewUrls(workspaceDto);
     }
 
     return workspaceDto;
   }
 
+  private void fixRuntimeCommandsPreviewUrls(WorkspaceDto workspaceDto) {
```

Comment:
I think this whole logic deserves to be in separate class (if it doesn't tied with service too much)